### PR TITLE
[HttpFoundation] Deprecate extending some methods

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/pull/19727 |
| License | MIT |
| Doc PR |  |

It's really hard to change methods signature because of bc. I'm proposing to deprecate extending some getters/setters of `Response` because of this (and because extending them is not really useful).
If you like this approach it could be used in other places to simplify bc in 4.0.

Edit: This causes issues (warnings always triggered) when mocking `Response` entirely but does it matter as people should only mock needed methods?

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 1]
```diff
@@ -201,6 +201,19 @@ public function __construct($content = '', $status = 200, $headers = array())
         $this->setContent($content);
         $this->setStatusCode($status);
         $this->setProtocolVersion('1.0');
+
+        // Deprecations
+        $class = get_class($this);
+        if (__CLASS__ === $class || 0 === strpos($class, 'Symfony\\')) {
```
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: maybe we should hardcode a list here to improve performance?

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: ok for me. what about this µoptim:?
`if (__CLASS__ === $class || 'S' !== $class[0] || 0 === strpos($class, 'Symfony\\Component\\')) {`

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 2]
```diff
@@ -201,6 +201,19 @@ public function __construct($content = '', $status = 200, $headers = array())
...
+            return;
+        }
+
+        foreach (array('setDate', 'setExpires', 'setLastModified', 'getDate', 'getExpires', 'getLastModified') as $method) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: array should be moved to a `private static`

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 3]
```diff
@@ -201,6 +201,19 @@ public function __construct($content = '', $status = 200, $headers = array())
...
+            $m = new \ReflectionMethod($this, $method);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: what about caching the result in a private static array?

File: src/Symfony/Component/HttpFoundation/Response.php (hunk 1 above)
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: Any vendor not beginning by `S` would pass the check. I'll try hard-coding a list.

File: src/Symfony/Component/HttpFoundation/Response.php (hunk 3 above)
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: good idea :+1: done

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 4]
```diff
@@ -186,6 +186,14 @@ class Response
         511 => 'Network Authentication Required',                             // RFC6585
     );
 
+    private static $deprecatedMethods = array('setDate', 'setExpires', 'setLastModified', 'getDate', 'getExpires', 'getLastModified');
+    private static $deprecationsTriggered = array(
+        __CLASS__ => true,
+        JsonResponse::class => true,
+        RedirectResponse::class => true,
+        StreamedResponse::class => true,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: BinaryFileResponse missing

File: src/Symfony/Component/Security/Http/Tests/Authentication/DefaultAuthenticationFailureHandlerTest.php [hunk 5]
```diff
@@ -47,7 +48,7 @@ public function testForward()
             ->method('createRequest')->with($this->request, '/login')
             ->will($this->returnValue($subRequest));
 
-        $response = $this->getMock('Symfony\Component\HttpFoundation\Response');
+        $response = new Response();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Tests/Authentication/DefaultAuthenticationFailureHandlerTest.php]: these changes are unrelated and should be done on 2.7, isn't it?

File: src/Symfony/Component/HttpFoundation/Response.php (hunk 4 above)
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: good catch, thanks !

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 6]
```diff
@@ -186,6 +186,16 @@ class Response
         511 => 'Network Authentication Required',                             // RFC6585
     );
 
+    private static $deprecatedMethods = array('setDate', 'setExpires', 'setLastModified', 'getDate', 'getExpires', 'getLastModified');
+    private static $deprecationsTriggered = array(
+        __CLASS__ => true,
+        ApacheRequest::class => true,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: this is Request, not Response :)

File: src/Symfony/Component/Security/Http/Tests/Authentication/DefaultAuthenticationFailureHandlerTest.php (hunk 5 above)
Comment (Ener-Getick) [inline src/Symfony/Component/Security/Http/Tests/Authentication/DefaultAuthenticationFailureHandlerTest.php]: Tests are failing without it (see https://ci.appveyor.com/project/fabpot/symfony/build/1.0.12032) but i'll remove them as soon as https://github.com/symfony/symfony/issues/19736 is merged.

File: src/Symfony/Component/HttpFoundation/Response.php (hunk 6 above)
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: Indeed, I really need an autocompleter... :P

Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: Well in fact that's request, it just shouldn't be here... ^^

**Comment by stof** [comment]
Regarding mocks, could we skip the warning for common mocking libraries ? At least PHPunit_MockObject and Prophecy are easy to detect, as their doubles are implementing an interface (``\PHPUnit_Framework_MockObject_MockObject`` for PHPUnit and ``\Prophecy\Doubler\DoubleInterface`` for Prophecy). I don't know Mockery, but it may be the same.

This would avoid causing issues for people mocking the Response class for other methods (Prophecy does not allow doing partial mocking for instance).

**Comment by Ener-Getick** [comment]
@stof fixed using `get_parent_class`. For Mockery it looks like it isn't extending the original class so it won't change anything.

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 7]
```diff
@@ -201,6 +210,23 @@ public function __construct($content = '', $status = 200, $headers = array())
...
+        if ($this instanceof \PHPUnit_Framework_MockObject_MockObject || $this instanceof \Prophecy\Doubler\DoubleInterface) {
+            $class = get_parent_class($class);
+        }
+        if (isset(self::$deprecationsTriggered[$class])) {
+            return;
+        }
+
+        self::$deprecationsTriggered[$class] = true;
+        foreach (self::$deprecatedMethods as $method) {
+            $m = new \ReflectionMethod($class, $method);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: $m should be $r "as usual"

Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: `$r` isn't for `ReflectionClass`?

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: I'd be $c then :) $r is for Reflector, the base interface

Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: Indeed makes sense now ^^

**Comment by nicolas-grekas** [comment]
rebase unlocked

**Comment by Ener-Getick** [comment]
@nicolas-grekas i already removed these commits.

**Comment by javiereguiluz** [comment]
@Ener-Getick could we estimate the performance hit related to the code added in the `Response` class (if any)? Thanks!

**Comment by Ener-Getick** [comment]
@javiereguiluz that's hard to benchmark but when I run `ResponseTest` in different conditions, I have no significant results (maybe a difference of 1 or 2ms and it originally took ~50ms).

**Comment by nicolas-grekas** [comment]
:+1: nice trick for eventually fixing these type hints in 4.0.
performance shouldn't be an issue: the new code is quick, and Response aren't instantiated that often.

**Comment by Ener-Getick** [comment]
At the same time, would you like to make `final` other methods of `Response`?

I think we can make final those methods:
- `setContent`/`getContent`
- `setProtocolVersion`/`getProtocolVersion`
- `setStatusCode`/`getStatusCode`
- `setCharset`/`getCharset`
- `setPrivate`/`setPublic`
- `getDate`/`setDate`
- `getAge`
- `getExpires`/`setExpires`
- `getMaxAge`/`setMaxAge`
- `setSharedMaxAge`
- `getTtl`/`setTtl`
- `setClientTtl`
- `getLastModified`/`setLastModified`
- `getEtag`/`setEtag`
- `hasVary`/`getVary`/`setVary`
- `isInvalid`/`isSuccessful`/`isSuccessful`/`isClientError`/`isServerError`
- `isOk`/`isForbidden`/`isNotFound`/`isRedirect`/`isEmpty`

Or is that too much? I think we can deprecate extending most methods and see if people complain about some deprecations. It won't be break any applications and would make bc easier if this methods aren't actually used (which should be the case as their content is very generic).

**Comment by fabpot** [comment]
The Response object is a data object that should be very "stable" (as HTTP is stable). So, I would say that extending most of the methods do not make sense indeed... if someone needs to do that, it's probably a bug that we should fix instead.

**Comment by Ener-Getick** [comment]
Ok, done.
The following methods will still be extendable (if you think we should also deprecate extensing some of them):
- `create`
- `__clone`
- `prepare`
- `sendHeaders`
- `sendContent`
- `send`
- `isCacheable`
- `isFresh`
- `isValidateable`
- `mustRevalidate`
- `expire`
- `setCache`
- `setNotModified`
- `isNotModified`
- `closeOutputBuffers`
- `ensureIEOverSSLCompatibility`

**Comment by jakzal** [comment]
@Ener-Getick `setContent()` is also [overridden by laravel](https://github.com/illuminate/http/blob/master/Response.php#L28). Before proceeding with this PR, we should check if there are any popular libraries extending methods you propose to make final.

**Comment by Ener-Getick** [comment]
Sure i don't think i'll have the time this week-end but maybe next week.

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 8]
```diff
@@ -186,6 +186,15 @@ class Response
         511 => 'Network Authentication Required',                             // RFC6585
     );
 
+    private static $deprecatedMethods = array('setDate', 'getDate', 'setExpires', 'getExpires', 'setLastModified', 'getLastModified', 'setProtocolVersion', 'getProtocolVersion', 'setStatusCode', 'getStatusCode', 'setCharset', 'getCharset', 'setPrivate', 'setPublic', 'getAge', 'getMaxAge', 'setMaxAge', 'setSharedMaxAge', 'getTtl', 'setTtl', 'setClientTtl', 'getEtag', 'setEtag', 'hasVary', 'getVary', 'setVary', 'isInvalid', 'isSuccessful', 'isRedirection', 'isClientError', 'isOk', 'isForbidden', 'isNotFound', 'isRedirect', 'isEmpty');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/Response.php]: could you put this array on several lines to ease reading please?

**Comment by nicolas-grekas** [comment]
I'm :+1: with the updated list

File: src/Symfony/Component/HttpFoundation/Response.php (hunk 8 above)
Comment (Ener-Getick) [inline src/Symfony/Component/HttpFoundation/Response.php]: :+1: done

**Comment by fabpot** [comment]
Thank you @Ener-Getick.

**Comment by fabpot** [comment]
@Ener-Getick What about doing the same with the Request class?

**Comment by Ener-Getick** [comment]
I have to take a look at it to see if that's worth it, i'll do that in a few days.

**Comment by Ener-Getick** [comment]
As a side node I searched methods extended in drupal and laravel and I only found `setContent` and `sendContent`, so no problem with this PR.
